### PR TITLE
[CI] Try introducing isort. 

### PR DESCRIPTION
**Pull request by rkooo567**
It is WIP, and Idk if committers agree on introducing this feature yet. 

Seems like vllm oss has sorting rule that's implicit. This PR tries introducing isort to `format.sh`. 

This PR only touches small # of files to avoid huge merge conflict. 

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by simon-mo** [comment]
Overall I'm in favor. But I think ruff already does this we just need to remove the ignore https://github.com/astral-sh/ruff/issues/8926#issuecomment-1834048218

**Comment by zhuohan123** [comment]
This is great! I'm thinking about having this for a long time.

**Comment by rkooo567** [comment]
@simon-mo I can also use ruff if that's what you prefer, so plz lmk.

# isort policy
The default ordering; 
```
FUTURE,STDLIB,THIRDPARTY,FIRSTPARTY,LOCALFOLDER
```
Also it prioritizes regular import over from import. I believe this is exact same policy as vllm, but please review ^ @simon-mo @zhuohan123 

# Merge plan
In terms of sorting all files, I will do incrementally to avoid merge conflict.

# CI
@simon-mo do Can you point me which file I have to modify to introduce sorting check in the CI?

**Comment by zhuohan123** [comment]
I think enabling ruff formatter is a big change and will probably not be done in the near future. Having isort as a temp fix is better.

**Comment by zhuohan123** [comment]
Also the order looks good to me!

File: vllm/config.py [hunk 1]
```diff
@@ -103,7 +103,8 @@ def __init__(
         if os.environ.get("VLLM_USE_MODELSCOPE", "False").lower() == "true":
             # download model from ModelScope hub,
             # lazy import so that modelscope is not required for normal use.
-            from modelscope.hub.snapshot_download import snapshot_download  # pylint: disable=C
+            from modelscope.hub.snapshot_download import \
+                snapshot_download  # pylint: disable=C
```
Comment (zhuohan123) [inline vllm/config.py:107]: Can we somehow config isort to use `()` for line continuation?
```suggestion
            from modelscope.hub.snapshot_download import (
                snapshot_download)  # pylint: disable=C
```

Comment (rkooo567) [reply vllm/config.py:107]: Yep. Will fix it by EoD today

**Comment by simon-mo** [comment]
Checkout .github folder for workflows for lint.

File: vllm/config.py (hunk 1 above)
Comment (rkooo567) [reply vllm/config.py]: fixed

**Comment by rkooo567** [comment]
@simon-mo can you take a look? I included this to a part of ruff btw (instead of adding a new step)

**Comment by rkooo567** [comment]
Looks like I need to make every file to pass isort to pass ruff. 

Let's merge this PR without ruff check. And then in the next PR, I am going to apply isort on all files + ruff check. And then let's merge fast before merge conflict happes?

**Comment by simon-mo** [comment]
Can you clarify the relationship here? I would assume once you enable the ruff check, isort is no longer needed. Supposedly developers can just run `ruff --check fix` to sort the imports

**Comment by rkooo567** [comment]
ah my wording was confusing. Sorry about that. More concretely;

- I added isort . --check-only to the end of ruff.yaml (so it will be included in the ci). See https://github.com/vllm-project/vllm/pull/3495/commits/c4681b02d0a06d7da4cb3096ee4c25376c97c1c5
- this fail because not every file is sorted yet.
- So my proposal is to merge this PR, and I make another PR to sort every file.

Alternatively, we can only check diff, but when I tried this, it somehow didin't work

**Comment by simon-mo** [comment]
Thanks for clarifying. My question is that, instead of using `isort` executable, can we just use `ruff` executable by uncomment this line 
https://github.com/vllm-project/vllm/blob/e90fc21f2eda7e53f692398ee2c0cb5a0ac19693/pyproject.toml#L29-L30

**Comment by rkooo567** [comment]
@simon-mo let me try that out!

File: .github/workflows/ruff.yml [hunk 2]
```diff
@@ -25,10 +25,10 @@ jobs:
...
       run: |
         python -m pip install --upgrade pip
-        pip install ruff==0.1.5 codespell==2.2.6 tomli==2.0.1
+        pip install ruff==0.1.5 codespell==2.2.6 tomli==2.0.1 isort==5.13.2
     - name: Analysing the code with ruff
       run: |
         ruff .
     - name: Spelling check with codespell
       run: |
-        codespell --toml pyproject.toml
```
Comment (zhuohan123) [inline .github/workflows/ruff.yml:34]: Nit: newline

File: pyproject.toml [hunk 3]
```diff
@@ -51,3 +51,7 @@ exclude = "vllm/model_executor/parallel_utils/|vllm/model_executor/models/"
 [tool.codespell]
 ignore-words-list = "dout, te, indicies"
 skip = "./tests/prompts"
+
+[tool.isort]
+use_parentheses = true
+skip_gitignore = true
```
Comment (zhuohan123) [inline pyproject.toml:57]: Nit: newline

**Review by zhuohan123 — APPROVED** [review]
Thanks for introducing this! Please let us know when this is ready to be merged. I personally also feel that we can apply isort to all files in this PR.

**Comment by rkooo567** [comment]
@simon-mo @zhuohan123 

- Tried ruff, and it doesn't support `use_parentheses`, so I just sticked to isort.
- All files are isorted.

**Pull request merged**